locale.c

```diff
@@ -1401,114 +1595,48 @@ S_setlocale_from_aggregate_LC_ALL(pTHX_ const char * locale, const line_t line)
 
     PERL_ARGS_ASSERT_SETLOCALE_FROM_AGGREGATE_LC_ALL;
 
-    /* If the string that gives what to set doesn't include all categories,
-     * the omitted ones get set to "C".  To get this behavior, first set
-     * all the individual categories to "C", and override the furnished
-     * ones below.  FALSE => No need to recalculate LC_ALL, as this is a
-     * temporary state */
-    if (! bool_setlocale_2008_i(LC_ALL_INDEX_, "C", line)) {
-        setlocale_failure_panic_c(LC_ALL, locale_on_entry, "C", __LINE__, line);
-        NOT_REACHED; /* NOTREACHED */
-    }
-
-    const char * s = locale;
-    const char * e = locale + strlen(locale);
-    while (s < e) {
-        const char * p = s;
-
-        /* Parse through the category */
-        while (isWORDCHAR(*p)) {
-            p++;
-        }
-
-        const char * category_end = p;
-
-        if (*p++ != '=') {
-            locale_panic_(Perl_form(aTHX_
-                          "Unexpected character in locale category name '%s"
-                          "<-- HERE",
-                          get_displayable_string(s, p - 1, 0)));
-        }
-
-        /* Parse through the locale name */
-        const char * name_start = p;
-        while (p < e && *p != ';') {
-            p++;
-        }
-        if (UNLIKELY( p < e && *p != ';')) {
-            locale_panic_(Perl_form(aTHX_
-                          "Unexpected character in locale name '%s<-- HERE",
-                          get_displayable_string(s, p, 0)));
-        }
+    const char * locale_categories[LOCALE_CATEGORIES_COUNT_];
+    switch (parse_LC_ALL_string(locale,
+                                (const char **) &locale_categories,
+                                line))
+    {
+      case invalid:
+        return NULL;
 
-        const char * name_end = p;
+      case no_array:
+        locale_panic_(Perl_form(aTHX_ "(%" LINE_Tf
+                                      "): expecting aggregate locale, got '%s'",
+                                      line, locale));
+        NOT_REACHED; /* NOTREACHED */
 
-        /* Space past the semi-colon */
-        if (p < e) {
-            p++;
-        }
+      case full_array:
+        break;
+    }
 
-        /* Find the index of the category name in our lists */
-        for (PERL_UINT_FAST8_T i = 0; i < LC_ALL_INDEX_; i++) {
+    /* Change each category to the value returned for it */
+    for (unsigned int i = 0; i < LC_ALL_INDEX_; i++) {
```

Comment:
I do prefer the more general types here.  While it's not relevant here (LC_ALL_INDEX_ is a small integer), in some other cases CodeQL complains about using the possibly-small-PERL_UINT_FAST8_T variable as a loop index when the loop limit is a bigger type, for example:
```
regexec.c:4952
        for (i = 0; i < folds_to_count
                      + UNLIKELY(multi_fold_from != 0); i++)
Comparison between i of type PERL_UINT_FAST8_T and ... + ... of wider type unsigned long.
```
(this is as https://github.com/tonycoz/perl5/security/code-scanning/27 which I'm not sure you can see)

Reply:
You can't increment an enum type in C++:
```
tony@venus:.../perl/git$ g++ -oenuminc enuminc.cpp 
enuminc.cpp: In function ‘int main()’:
enuminc.cpp:9:33: error: no match for ‘operator++’ (operand type is ‘foo’)
    9 |     for (foo i = a; i < nomore; ++i)
      |                                 ^~~
tony@venus:.../perl/git$ cat enuminc.cpp
#include <iostream>

enum foo {
    a, b, nomore
};

int
main() {
    for (foo i = a; i < nomore; ++i)
        std::cout << i << '\n';
}
```
unless you provide an overloaded operator:
```
tony@venus:.../perl/git$ g++ -oenuminc enuminc.cpp 
tony@venus:.../perl/git$ cat enuminc.cpp
#include <iostream>

enum foo {
    a, b, nomore
};

foo operator++(foo &x) {
    x = static_cast<foo>(x+1);
    return x;
}

int
main() {
    for (foo i = a; i < nomore; ++i)
        std::cout << i << '\n';
}
tony@venus:.../perl/git$ ./enuminc
0
1
```

Reply:
Right, but I think it is better to wait to discuss this until we get to a commit that actually might use this